### PR TITLE
Adding reset epoch protocol message and support

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #816

Differential Revision: [D40273860](https://our.internmc.facebook.com/intern/diff/D40273860)

### CODE REVIEW

File: torchdata/dataloader2/communication/iter.py [hunk 1]
```diff
@@ -123,7 +123,12 @@ def DataPipeBehindQueues(source_datapipe, protocol, full_stop=False, blocking_re
             yield True
             continue
 
-        if isinstance(request, communication.messages.ResetIteratorRequest):
+        if isinstance(request, communication.messages.ResetEpochRequest):
+            if reset_epoch_fn is not None:
+                reset_epoch_fn(source_datapipe, request.args)
```
Comment (ejguan) [inline torchdata/dataloader2/communication/iter.py:128]: Should we assume `args` as a list and do `*request.args`?

File: torchdata/dataloader2/communication/map.py [hunk 2]
```diff
@@ -100,7 +100,12 @@ def DataPipeBehindQueues(source_datapipe, protocol, full_stop=False, blocking_re
             yield True
             continue
 
-        if isinstance(request, communication.messages.TerminateRequest):
+        if isinstance(request, communication.messages.ResetEpochRequest):
+            if reset_epoch_fn is not None:
+                reset_epoch_fn(source_datapipe, request.args)
```
Comment (ejguan) [inline torchdata/dataloader2/communication/map.py:105]: Ditto

File: torchdata/dataloader2/reading_service.py [hunk 3]
```diff
@@ -187,10 +204,13 @@ def initialize(self, datapipe: DataPipe) -> DataPipe:
             )
             self.datapipes.append(local_datapipe)
 
-        return _IterateQueueDataPipes(self.datapipes)  # type: ignore[return-value]
+        self.datapipes_iterator = _IterateQueueDataPipes(self.datapipes)
```
Comment (ejguan) [inline torchdata/dataloader2/reading_service.py:207]: nit: Could we please rename it as it's still a `DataPipe` not an `Iterator`?

**Review by ejguan — APPROVED** [review]
Thank you for the quick implementation!

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40273860).

File: torchdata/dataloader2/communication/iter.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torchdata/dataloader2/communication/iter.py]: Done

File: torchdata/dataloader2/communication/map.py (hunk 2 above)
Comment (VitalyFedyunin) [reply torchdata/dataloader2/communication/map.py]: Done

File: torchdata/dataloader2/reading_service.py (hunk 3 above)
Comment (VitalyFedyunin) [reply torchdata/dataloader2/reading_service.py]: Done

**Review by NivekT — APPROVED** [review]
LGTM.

Question:
Is the main difference between `ResetEpoch` and `ResetIterator` the fact that the new one will accepts a seed (or other arguments required) to restart the epoch?

**Comment by ejguan** [comment]
> Is the main difference between `ResetEpoch` and `ResetIterator` the fact that the new one will accepts a seed (or other arguments required) to restart the epoch?

That's actually a good question. We might need to change the message name from `ResetIterator` to `InitIterator`.

**Pull request closed without merging**